### PR TITLE
Validate the undefined parts in a `Config.toml` 

### LINKED ISSUE #28732: Identify and validate unwanted TOML parts in `Config.toml`
**Description:**
Currently, if we have any extra content that has a valid TOML format inside the `Config.toml` file, it is ignored in the configurable implementation. It is preferred to throw error if such unnecessary content is given in the TOML file. 

**Steps to reproduce:**
Ballerina context in module `main`
```ballerina
configurable int x = ?;

```
In Config.toml,

```toml
[main]
x = 11
y = "abc"  #No configurable variable with name 'y'

[extra]
z = "def" #No module named 'extra' and no variable named 'z'
```

**Affected Versions:**
alpha2

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #30160: Gets a warning when there exists an empty Config.toml file
Why do we issue a warning when there exists an empty `Config.toml` file in the current working directory? This is unnecessary IMO. 

```
warning: an empty configuration file is found in path '/Users/...../Config.toml'. Please provide values for configurable variables
```

### LINKED ISSUE #30163: An unnecessary error is reported when a value is not supplied for a mandatory configurable variable in a sub module
Say there exists a submodule in a package, and that module contains a mandatory configurable variable. When a value is not supplied via `Config.toml`, I get the following errors. The second error is misleading and unnecessary

```
error: value not provided for required configurable variable 'name'
error: [Config.toml:(1:1,11:16)] invalid module structure found for module 'config1.log'. Please provide the module name as '[config1.log]'
```

My `Config.toml` does not contain any configuration values related to the submodule.

### LINKED ISSUE #30164: Configurable related error messages do not required contextual information
Case 1)
When I value is not provided for a mandatory configurable variable, I get the following error. It doesn't say anything about the module name or the package name.

```
error: value not provided for required configurable variable 'name'
```

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
Validate the undefined parts in a `Config.toml` file. 
Fixes #28732 
Fixes #30163 
Fixes #30164 
Fixes #30160 

## Approach
The non-visited TOML parts are identified using a `Set` that keeps all visited parts. 
- It gives a warning if there's a undefined part found in the `Config.toml`. 
- The `invalid module structure` warning is given when an imported module that contains all optional variables is given without it's org-name.  

```toml
mainVar = "variable from root module"
undefinedVar = "variable not defined"

[undefined_module]
testVar1 = "variable from an undefined module"

[[undefined_table]]
testVar2 = "table type variable from an undefined module"
```
The warning will be,
```
warning: [Config.toml:(1:1,1:38)] unused configuration value 'mainVar'
warning: [Config.toml:(2:1,2:38)] unused configuration value 'undefinedVar'
warning: [Config.toml:(4:1,5:47)] unused configuration value 'undefined_module'
warning: [Config.toml:(5:1,5:47)] unused configuration value 'undefined_module.testVar1'
warning: [Config.toml:(7:1,8:58)] unused configuration value 'undefined_table'
warning: [Config.toml:(8:1,8:58)] unused configuration value 'undefined_table.testVar2'
```

- Removed redundant warnings for empty toml file and empty toml string
- Modified the module toml structure error to the following message
```
[Config.toml:(2:1,2:21)] invalid toml structure found for module 'test_module.foo'. Please provide the module name as '[test_module.foo]'
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by manuranga** [comment]
I find undefined to be confusing for the person who is running, specially if the person who is writing the toml is different form the person who is writing the program. Saying `extra`, `unused` or `surplus` makes more sense in that POV.

How about something like below :
```
warning: 
[Config.toml:(2:1,2:38)] Unused configuration value 'undefinedVar'
[Config.toml:(4:1,5:47)] Unused configuration value 'undefined_module'
[Config.toml:(5:1,5:47)] Unused configuration value 'undefined_module.testVar1'
[Config.toml:(7:1,8:58)] Unused configuration value 'undefined_table'
[Config.toml:(8:1,8:58)] Unused configuration value 'undefined_table.testVar2'

```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/ModuleInfo.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    void analyseModules(Module rootModule) {
+        for (Module entry : moduleSet) {
+            String rootModuleName = rootModule.getName();
+            if (rootModuleName.startsWith(entry.getOrg())) {
+                hasModuleAmbiguity = true;
+            }
+            String moduleName = entry.getName();
+            if (moduleName.startsWith(rootModuleName + ".")) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/ModuleInfo.java:51]: Will this log work if my root module is a.b and my module name is c.a.b.e

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java [hunk 2]
```diff
@@ -104,12 +107,76 @@
...
+        for (Map.Entry<String, TopLevelNode> nodeEntry : baseNode.entries().entrySet()) {
+            TomlNode node = nodeEntry.getValue();
+            String lineRange = getLineRange(node);
+            String entryKey = nodeEntry.getKey();
+            boolean isInvalidNode = invalidTomlLines.contains(node.location().lineRange());
+            switch (node.kind()) {
+                case KEY_VALUE:
+                    if (!visitedNodes.contains(node) && !isInvalidNode) {
+                        exceptionList
+                                .add(new ConfigException(CONFIG_TOML_UNUSED_VALUE, lineRange, moduleHeader + entryKey));
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java:139]: Creating a list of exceptions and added to the list, in this case, is not nice. Shall we change the complete method to pass the diagnostics log?
 void complete(RuntimeDiagnosticLog diagnosticLog);

Comment (HindujaB) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java:139]: +1

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**